src/goto-instrument/contracts/dynamic-frames/dfcc_loop_utils.cpp

```diff
@@ -0,0 +1,514 @@
+const irep_idt ID_loop_instr_type = "loop-instr-type";
+const irep_idt ID_loop_instr_body = "body";
+const irep_idt ID_loop_instr_head = "head";
+const irep_idt ID_loop_instr_latch = "latch";
+const irep_idt ID_loop_instr_exiting = "exiting";
+
+void tag_loop_instr(
+  goto_programt::instructiont::targett &target,
+  const irep_idt &instr_type,
+  const std::size_t loop_nesting_id)
+{
+  irept &irep = target->source_location_nonconst();
+  irep.set(ID_loop_instr_type, instr_type);
+  irep.set(ID_loop_nesting_id, loop_nesting_id);
+}
+
+irep_idt get_loop_instr_type(const goto_programt::instructiont::targett &target)
+{
+  irept &irep = target->source_location_nonconst();
+  return irep.get(ID_loop_instr_type);
+}
+
+std::size_t
+get_loop_nesting_id(const goto_programt::instructiont::targett &target)
+{
+  PRECONDITION(!get_loop_instr_type(target).empty());
+  irept &irep = target->source_location_nonconst();
+  return irep.get_size_t(ID_loop_nesting_id);
+}
+
+bool is_loop_instr(
+  const goto_programt::instructiont::targett &target,
+  std::size_t loop_nesting_id)
+{
+  if(get_loop_instr_type(target).empty())
+    return false;
+  return get_loop_nesting_id(target) == loop_nesting_id;
+}
+
+bool is_loop_exiting(const goto_programt::instructiont::targett &target)
+{
+  return get_loop_instr_type(target) == ID_loop_instr_exiting;
+}
+
+bool is_loop_head(const goto_programt::instructiont::targett &target)
+{
+  return get_loop_instr_type(target) == ID_loop_instr_head;
+}
+
+void remove_loop_instr_type(source_locationt &source_location)
+{
+  source_location.remove(ID_loop_instr_type);
+  source_location.remove(ID_loop_nesting_id);
+}
+
+void remove_loop_instr_type(goto_programt::targett &target)
+{
+  if(!get_loop_instr_type(target).empty())
+  {
+    target->source_location_nonconst().remove(ID_loop_instr_type);
+    target->source_location_nonconst().remove(ID_loop_nesting_id);
+
+    return;
+  }
+}
+
+void remove_loop_instr_type(goto_programt &function_body)
+{
+  for(auto i = function_body.instructions.begin();
+      i != function_body.instructions.end();
+      i++)
+  {
+    remove_loop_instr_type(i);
+  }
+}
+
+void preprocess_and_check_loops(goto_programt &function_body, messaget &log)
+{
+  natural_loops_mutablet natural_loops(function_body);
+
+  for(auto &loop_head_and_content : natural_loops.loop_map)
+  {
+    auto &loop_content = loop_head_and_content.second;
+    if(loop_content.empty())
+      continue;
+
+    auto loop_head = loop_head_and_content.first;
+    auto loop_end = loop_head;
+
+    // Find the last back edge to `loop_head`
+    for(const auto &t : loop_content)
+    {
+      if(
+        t->is_goto() && t->get_target() == loop_head &&
+        t->location_number > loop_end->location_number)
+        loop_end = t;
+    }
+
+    if(loop_end == loop_head)
+    {
+      log.error() << "Could not find end of the loop starting at: "
+                  << loop_head->source_location() << messaget::eom;
+      throw 0;
+    }
```

Comment:
Should this be a `DATA_INVARIANT`? We throw exceptions when something might be actionable for the user, but I'm not sure what the user could do here?